src/reason-parser/reason_parser.mly

```diff
@@ -608,30 +608,96 @@ let process_underscore_application args =
-     * f((.) => 42, 1000) -> f([@bs] () => 42, 1000) (attr needs to stay on the callback) *)
-    | (Nolabel, ({pexp_attributes; pexp_desc} as e))::tl when
-          match pexp_desc with
-          | Pexp_fun _ -> false
-          | _ -> true
-      ->
-      let (new_attrs, uncurried) = match pexp_attributes with
-      | ({txt = "bs"}, PStr [])::tl -> (tl, true)
-      | attrs -> (attrs, false)
-      in
-      (((Nolabel, {e with pexp_attributes = new_attrs})::tl), uncurried)
-  | xs -> (xs, false)
-  end in
   let loc = mklocation startp endp in
-  let attrs = if uncurried then [uncurry_payload loc] else [] in
-  let (args, wrap) = process_underscore_application args in
-  if args = [] then { body with pexp_loc = loc } else
-    wrap (mkexp ~attrs ~loc (Pexp_apply (body, args)))
+  if args = [] then {body with pexp_loc = loc}
+  else
+  (*
+   * Post process the arguments and transform [@uncurry] into [@bs].
+   * Returns a tuple with a boolean (was it uncurried?) and
+   * the posible rewritten arg.
+   *)
+  let rec process_args acc es =
+    match es with
+    | (lbl, e)::es ->
+        let attrs = e.pexp_attributes in
+        let hasUncurryAttr = ref false in
+        let newAttrs = List.filter (function
+          | ({txt = "uncurry"}, PStr []) ->
+              hasUncurryAttr := true;
+              false
+          | _ -> true) attrs
+        in
+        let uncurried = !hasUncurryAttr in
+        let newArg = (lbl, { e with pexp_attributes = newAttrs }) in
+        process_args ((uncurried, newArg)::acc) es
+    | [] -> acc
+    in
+    (*
+     * Groups all uncurried args falling under the same Pexp_apply
+     * Example:
+     *    add(. 2, 3, . 4, 5) or add(. 2, 3)(. 4, 5)  (equivalent)
+     * This results in two groups: (true, [2, 3]) & (true, [4, 5])
+     * Both groups have 'true' as their first tuple element, because
+     * they are uncurried.
+     * add(2, 3, . 4) results in the groups (false, [2, 3]) & (true, [4])
+     *)
+    let rec group grp acc = function
+    | (uncurried, arg)::xs ->
+        let (_u, grp) = grp in
+        if uncurried = true then begin
+          group (true, [arg]) ((_u, (List.rev grp))::acc) xs
+        end else begin
+          group (_u, (arg::grp)) acc xs
+        end
+    | [] ->
+        let (_u, grp) = grp in
+        List.rev ((_u, (List.rev grp))::acc)
+    in
+    (*
+     * Recursively transforms all groups into a (possibly uncurried)
+     * Pexp_apply
+     *
+     * Example:
+     *   Given the groups (true, [2, 3]) & (true, [4, 5]) and body 'add',
+     *   we get the two nested Pexp_apply associated with
+     *   (add(. 2, 3))(. 4, 5)
+     *)
+    let rec make_appl body = function
+      | args::xs ->
+          let (uncurried, args) = args in
+          let expr = if args = [] then body
+          else
+            let (args, wrap) = process_underscore_application args in
+            let expr = mkexp (Pexp_apply (body, args)) in
```

Comment:
How about taking the location range from the arguments:
```
            let args_loc = match args, List.rev args with
              | ((_, s)::_), ((_, e)::_) -> mklocation s.pexp_loc.loc_start e.pexp_loc.loc_end
              | _ -> assert false in
            let expr = mkexp (Pexp_apply (body, args)) ~loc:args_loc in

```

Reply:
Thanks for the code, exactly what was needed.